### PR TITLE
Fix tab completions for a hash table

### LINKED ISSUE #5322: Completion of hash table key has invalid replacement index/length
Steps to reproduce
------------------

```powershell
$line = 'gps | format-table -Property @{'
$exp = TabExpansion2 $line $line.Length
$exp.ReplacementIndex
$exp.ReplacementLength
$exp.CompletionMatches.Count
```

Expected behavior
-----------------
ReplacementIndex should be `$line.Length`, the length should be 0.

```none
31
0
5
```
If you change the repro to add a space at the end, then you will see 32, 0, 5
```powershell
$line = 'gps | format-table -Property @{ '
```
Actual behavior
---------------
The replacement index and length are negative.

```none
-1
-1
5
```
This raises an exception in PSReadLine (https://github.com/lzybkr/PSReadLine/issues/534)
In Windows PowerShell, if you remove PSReadLine and try tab interactively (instead of manually calling `TabExpansion2`), PowerShell will crash - as in exit after reporting a crash to Watson.
PowerShell ISE doesn't crash, but it also doesn't complete anything.

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.1.16299.19                                                                            
PSEdition                      Desktop                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.16299.19                                                                           
CLRVersion                     4.0.30319.42000                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 



Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      6.0.0-beta.9                                                                            
PSEdition                      Core                                                                                    
GitCommitId                    v6.0.0-beta.9                                                                           
OS                             Microsoft Windows 10.0.16299                                                            
Platform                       Win32NT                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 
WSManStackVersion              3.0                                                                                     
```

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

Fix #5322.

Now working:
- Get-Date | Sort-Object @{Expression=\<tab>
- Get-Date | Sort-Object @{Expression=...;\<tab>

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 1]
```diff
@@ -473,6 +473,13 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
                         replacementLength = 0;
                         break;
 
+                    case TokenKind.Semi:
```
Comment (lzybkr) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Does adding this case affect `get-something; e<TAB>` - in other words, no hashtable, just simple command completion?

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: No, I don't see the side affect.

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: I re-check `get-something; e<Tab>` - in Windows PowerShell 5.1 and in PowerShell Core 6.0 simple command completion works - nothing new added in the PR.

**Comment by iSazonov** [comment]
@SteveL-MSFT @daxian-dbw Could you please review too?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 2]
```diff
@@ -473,6 +473,13 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
...
+                        // Handle scenarios such as 'gci | Format-Table @{Label=...;<tab>'
+                        result = GetResultForHashtable(completionContext);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: `get-something;<tab>` does go through this path. It's better to check `if (lastAst is HashtableAst)` before calling `GetResultForHashtable`.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 3]
```diff
@@ -531,6 +538,18 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
...
+                                if (lastAst.Parent is DynamicKeywordStatementAst)
+                                {
+                                    // Handle scenarios such as 'configuration foo { File ab { Attributes ='
+                                    bool unused;
+                                    result = GetResultForEnumPropertyValueOfDSCResource(completionContext, string.Empty, ref replacementIndex, ref replacementLength, out unused);
+                                    break;
+                                }
+                                // Handle scenarios such as 'gci | Format-Table @{Label=<tab>'
+                                return null;
+                            }
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: This duplicates the call to `GetResultForEnumPropertyValueOfDSCResource`.
How about changing to this:
```
case TokenKind.LParen:
    {
        if (lastAst is AttributeAst)
        {
            ...
        }
        else if (lastAst.Parent is DynamicKeywordStatementAst)  // <-- change here
        {
            ...
        }
        else if (lastAst is HashtableAst hashTableAst && CheckForPendingAssignment(hashTableAst)) // <-- change here
        {
            return result;
        }
        break;
    }
```

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 [hunk 4]
```diff
@@ -65,4 +65,19 @@ Describe "Tab completion bug fix" -Tags "CI" {
...
+        $result.CompletionMatches.Count | Should -Be 0
+    }
+
+    It "Issue#5322 - 'Get-Date | Sort-Object @{Expression=...;' should work" {
+        $cmd = "Get-Date | Sort-Object @{Expression=...;"
+        $result = TabExpansion2 -inputScript $cmd -cursorColumn $cmd.Length
+        $result.CompletionMatches.Count | Should -Be 3
+        $result.CompletionMatches[0].CompletionText | Should -Be 'Expression'
+        $result.CompletionMatches[1].CompletionText | Should -Be 'Ascending'
+        $result.CompletionMatches[2].CompletionText | Should -Be 'Descending'
```
Comment (daxian-dbw) [inline test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: This test succeeds even without changes in this PR. You need to check `ReplacementIndex` and `ReplacementLength`, as they were -1 previously and changes in this PR should fix them.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Second condition doesn't work for `configuration foo { File ab { Attributes =(` test - really `lastAst.Parent` is `CommandExpressionAst`.

Fixed.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Fixed.

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 (hunk 4 above)
Comment (iSazonov) [reply test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: Fixed.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 5]
```diff
@@ -537,10 +547,10 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
                                 completionContext.ReplacementLength = replacementLength = 0;
                                 result = GetResultForAttributeArgument(completionContext, ref replacementIndex, ref replacementLength);
                             }
-                            else
+                            else if (lastAst.Parent is DynamicKeywordStatementAst || lastAst.Parent is CommandExpressionAst)
                             {
-                                //
-                                // Handle auto completion for enum/dependson property of DSC resource,
+                                // 1. Handle scenarios such as 'configuration foo { File ab { Attributes ='
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: > Handle auto completion for **enum/dependson** property of DSC resource

The example `configuration foo { File ab { Attributes =` is a completion for enum property, so it should not be marked as `1.`.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 6]
```diff
@@ -537,10 +547,10 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
                                 completionContext.ReplacementLength = replacementLength = 0;
                                 result = GetResultForAttributeArgument(completionContext, ref replacementIndex, ref replacementLength);
                             }
-                            else
+                            else if (lastAst.Parent is DynamicKeywordStatementAst || lastAst.Parent is CommandExpressionAst)
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: We are making two special checks here. Maybe we should only have one special check, like the following. What do you think?

```c#
if (lastAst is AttributeAst)
{
    // the original attribute handling
}
else if (lastAst is HashtableAst hashTableAst &&
         !(lastAst.Parent is DynamicKeywordStatementAst) &&
         CheckForPendingAssignment(hashTableAst))
{
    return null;
}
else
{
    // the original else block
}
```

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 [hunk 7]
```diff
@@ -65,4 +65,21 @@ Describe "Tab completion bug fix" -Tags "CI" {
...
+        $cmd = "Get-Date | Sort-Object @{Expression="
+        $result = TabExpansion2 -inputScript $cmd -cursorColumn $cmd.Length
+        $result.CompletionMatches.Count | Should -Be 0
+    }
+
+    It "Issue#5322 - 'Get-Date | Sort-Object @{Expression=...;' should work" {
+        $cmd = "Get-Date | Sort-Object @{Expression=...;"
+        $result = TabExpansion2 -inputScript $cmd -cursorColumn $cmd.Length
+        $result.CurrentMatchIndex | Should -Be -1
+        $result.ReplacementIndex | Should -Be 40
```
Comment (daxian-dbw) [inline test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: And `ReplacementLength`?

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 [hunk 8]
```diff
@@ -65,4 +65,21 @@ Describe "Tab completion bug fix" -Tags "CI" {
...
+        $result.CompletionMatches[2].CompletionText | Should -Be 'Descending'
+        $result.CompletionMatches[0].CompletionText | Should -Be 'Expression'
+        $result.CompletionMatches[1].CompletionText | Should -Be 'Ascending'
+        $result.CompletionMatches[2].CompletionText | Should -Be 'Descending'
```
Comment (daxian-dbw) [inline test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: `$result.CompletionMatches[2].CompletionText` is checked twice.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Fixed.

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 (hunk 7 above)
Comment (iSazonov) [reply test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: Fixed.

File: test/powershell/Host/TabCompletion/BugFix.Tests.ps1 (hunk 8 above)
Comment (iSazonov) [reply test/powershell/Host/TabCompletion/BugFix.Tests.ps1]: It was for ReplacementLength :-)

Fixed.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Looks good. Thanks!

Fixed.

**Comment by iSazonov** [comment]
The while code is very tricky and sensetive :confused: and crush PowerShell if something wrong. Maybe we should make it more robust.

**Review by daxian-dbw — APPROVED** [review]
LGTM.
@iSazonov What `while` code were you referring to?

**Comment by iSazonov** [comment]
I means our Completion code - if I return wrong index PowerShell is crashed. Asts can be combined in many ways. Can we be sure that we tested all their? The code is complex and tricky - can we create simple algorithm? Using visitors?

**Comment by iSazonov** [comment]
@lzybkr @SteveL-MSFT Do you have any comments?

**Comment by daxian-dbw** [comment]
@iSazonov Sorry for the late reply. I agree the tab completion code is a bit hard to organize now. Many changes have been made that are very specific to the scenarios they try to solve, so you can see checks in `CompletionAnalysis` that are tied to one particular scenario which is hard to understand just by reading the code.

However, the crash is possibly not related to tab completion, but the code rending results to console. If you mess up the index, the rendering code might do weird bad things. I don't think a visitor can help much here, as many checks need to inspect ancestors of an ast.

**Comment by iSazonov** [comment]
@daxian-dbw Thanks for great comment!

**Pull request merged**